### ISSUE TITLE
Provide APIs to 3p extensions for analytics custom events 

### ISSUE THREAD

**Issue by zhouyx** [ampproject/amphtml#9484]
Two API methods are provided for extension element to send analytics pings and customize their pings in different use cases.

### Manual trigger - CustomEventReporter class
Analytics triggers typically come from user events but one can also do manual triggering to send analytics pings with their customized data. 

To trigger custom events, 3P extension element first need to build a `customEventReporter` using `customEventReporterBuilder`.

For example
```
  buildCallback() {
   const builder = new customEventReporterBuilder(this.element);
   builder.track('layout', 'example.come/layout');
   builder.track('id', ['example.com/${id}', 'example2.com/${id}']);
   this.reporter_ = builder.build();
  }
```
`customEventReporterBuilder` constructor requires the parent AMP element. A builder instance provide two methods `track(eventName, request)` and `build()`. `#track()` takes in a custom event name, and the corresponding request (or an array of requests) to that event. While `#build()` should be called after adding all custom event, and will return the built `customEventReporter` instance.

The `customEventReporter` class provide a `trigger` method to manual trigger created custom events on 3P extension element with optional vars. 

For example
```
layoutCallback() {
  this.reporter_.trigger('layout');
  getIdPromise.then(id => {
    this.reporter_.trigger('id', {'id': '123'});
  }
}
```

Please note that with the use of `CustomEventReporter`, we assume all custom events are sandbox events that are scoped inside extension element. Thus `CustomEventReporter` can only used to report and collect 3P extension element created custom events. 

### Config 
Extensions can configure their [analytics config](https://github.com/ampproject/amphtml/blob/master/extensions/amp-analytics/amp-analytics.md#configuration) to get analytics events.

- To create the adapter class, extension element needs to provide a function that returns a promise with analytics config. 
```
const adapter = new SandboxAnalyticsAdapter(this.element, createConfigFunc);
```
The analytics config can be static, and checked in to the extension folder. For example:
```
import {analyticsConfig} from 'extensions/amp-***/0.1/analytics-config.js;
createConfigFunc = () => return Promise.resolve(analyticsConfig);
```
Or the analytics config can be fetched remotely, and available asynchronously.  For example:
```
createConfigFunc = fetchRemoteInfo.then(info => {
   return extractAnalyticsConfig(info);
};
```
- Restrictions apply to sandboxed analytics element's config. These Includes not supporting "type" or "config" attributes, limiting selector in parent element's scope, and applying variable replacement for whitelisted variables only.

### Behind the scene

`SandboxAnalyticsAdapter` will insert an `<amp-analytics>`  to DOM when its parent extension element starts layout. We delay the creation of analytics element because it's rare that an extension element collect analytics data before layout. And analytics is by default given lower priority than other contents. 

Those inserted amp-analytics elements are sandboxed in its parent extension element. This means it cannot access element outside its parent scope, and the [url variable substitution](https://github.com/ampproject/amphtml/blob/master/spec/amp-var-substitutions.md) only works if variable is included in the [whitelist](https://github.com/ampproject/amphtml/blob/master/extensions/amp-analytics/0.1/sandbox-vars-whitelist.js). 

The inserted `<amp-analytics>` element is not ready to handle analytics events until it has been initialized.  Before that `SandboxAnalyticsAdapter` class will buffer manual triggered events.  Please also note that the buffer has a size limit, and it only works for manual triggered events, not for other events happen before layout. 

If the extension doesn't [require relayout after unlayout](https://github.com/ampproject/amphtml/blob/master/src/base-element.js#L447), one can ignore this part. 
In the case of relayout, the adapter class will clean old buffered events, and wait for the new analytics element's ready signal. It is possible that elements can get laid out multiple times. One need to make sure the `createConfigFunc` returns the new  promise with new config after relayout.

**Comment by zhouyx** [comment]
#9506

**Comment by zhouyx** [comment]
@rudygalfi @lannka please take a quick look at the design. I'll work on it once everyone thinks it looks good. 
Once the implementation is done we should document it somewhere for extensions developers to use amp analytics.

**Comment by rudygalfi** [comment]
High-level looks OK, but it's a bit hard to discern which areas need my closest review. Let me know if there's specific aspects I should take a closer look at.

**Comment by zhouyx** [comment]
@rudygalfi my apology that the doc goes into too much details. I changed it a bit. A closer look at the Config & Manual Trigger part would be good. Thanks!

**Comment by rudygalfi** [comment]
Thanks. What does "not supporting remote config fetching with url keyword" mean?

**Comment by zhouyx** [comment]
Sorry I was trying to say 'config' or 'type' keyword, where they use default type config or fetch remote config through a provided url. I changed the doc.

**Comment by rudygalfi** [comment]
@zhouyx Makes sense - thanks!
